compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt

```diff
@@ -122,6 +146,59 @@ class SwitchGenerator(private val expression: IrWhen, private val data: BlockInf
         return true
     }
 
+    private fun IrCall.getValueArgumentThroughCoercion(index: Int): IrExpression? {
+        if (symbol == codegen.context.ir.symbols.unsafeCoerceIntrinsic) {
+            val coercee = getValueArgument(0)!!
+            require(coercee is IrCall)
+            return coercee.getValueArgument(index)
+        } else {
+            return getValueArgument(index)
+        }
+    }
+
+    private fun <T> IrExpression.asThroughCoercion(): T {
+        if (this is IrCall && symbol == codegen.context.ir.symbols.unsafeCoerceIntrinsic) {
+            val coercee = getValueArgument(0)!!
+            return coercee as T
+        } else {
+            return this as T
+        }
+    }
+
+    // Extract the switch subject and calculate labels and expressions of conditions.
+    // Every condition is of the form:
+    //   CALL EQEQ(CALL <unsafe-coerce><UInt,Int>(tmp_variable),
+    //             CALL <unsafe-coerce><UInt,Int>(constant))
+    private fun extractUIntSwitchComponents(callToLabels: List<CallToLabel>, expressionToLabels: ArrayList<ExpressionToLabel>): Triple<IrGetValue, ArrayList<ExpressionToLabel>, List<ValueToLabel>> {
+        val subject = callToLabels[0].call.getValueArgumentThroughCoercion(0)!!.asThroughCoercion<IrGetValue>()
+        val cases = callToLabels.map {
+            ValueToLabel(
+                it.call.getValueArgumentThroughCoercion(1)!!.asThroughCoercion<IrConst<*>>().value,
+                it.label
+            )
+        }.distinctBy { it.value }
+        val reachableLabels = HashSet(cases.map { it.label })
+        expressionToLabels.removeIf { it.label !in reachableLabels }
+        return Triple(subject, expressionToLabels, cases)
+    }
+
+    private fun extractSwitchComponents(callToLabels: List<CallToLabel>, expressionToLabels: ArrayList<ExpressionToLabel>): Triple<IrGetValue, ArrayList<ExpressionToLabel>, List<ValueToLabel>> {
+        // Get the
+        val subject = callToLabels[0].call.getValueArgument(0)!! as IrGetValue
```

Comment:
It seems to me the only difference between `extractUIntSwitchComponents` and `extractSwitchComponents` are there is a call to `getValueArgumentThroughCoercion()` for the arguments. But since it falls back to `getValueArgument()` if the call symbol is not the unsafe-coerce, then you could just always call `getValueArgumentThroughCoercion()` in both cases. Basically, I'm suggesting we just keep one version and always call `getValueArgumentThroughCoercion()`.

Reply:
I went a different route and wrote out the UInt case explicitly as the exception. I feel it was more direct and, as you suggested elsewhere, made the expected IR structure clearer.